core/src/rest/http_client.ts

```diff
@@ -0,0 +1,240 @@
+ * While Bot token is used for authorizing a Bot User account.
+ */
+export enum TokenType {
+  Bearer = "Bearer",
+  Bot = "Bot",
+}
+
+export interface HTTPClientOptions {
+  /** Type of token used for authorization. */
+  tokenType?: TokenType;
+
+  /** Token to be used for authorization. */
+  token?: string;
+
+  /** The User-Agent value to set. Default is a library name and repo URL. */
+  userAgent?: string;
+
+  /**
+   * The base URL to use for requests.
+   *
+   * Default is exported as `DISCORD_BAPI_BASE` from `types/src/constants.ts`.
+   */
+  baseURL?: string;
+
+  /**
+   * The API version to use for requests.
+   *
+   * Default is exported as `DISCORD_API_VERSION` from `types/src/constants.ts`.
+   */
+  version?: number;
+
+  /** Max request retries. Defaults to 5. */
+  maxRetries?: number;
+}
+
+export class Queue {
+  #promises: {
+    wait: Promise<void>;
+    resolve: () => void;
+  }[] = [];
+
+  resetTime = 0;
+
+  get length() {
+    return this.#promises.length;
+  }
+
+  async wait() {
+    if (this.resetTime > Date.now()) {
+      await new Promise((r) => setTimeout(r, this.resetTime - Date.now()));
+    }
+
+    const next = (this.#promises[this.#promises.length - 1]?.wait ??
+      Promise.resolve());
+
+    let resolve!: () => void;
+    const wait = new Promise<void>((r) => (resolve = r));
+
+    this.#promises.push({
+      wait,
+      resolve,
+    });
+
+    await next;
+  }
+
+  shift() {
+    let item;
+    if ((item = this.#promises.shift())) {
+      item.resolve();
+    }
+  }
+}
+
+/**
+ * HTTP Client manages requests to be sent to Discord REST API.
+ *
+ * This client manages rate limits, authorization, retries, etc.
+ */
+export class HTTPClient implements HTTPClientOptions {
+  tokenType?: TokenType;
+  token?: string;
+  userAgent: string;
+  baseURL: string;
+  version: number;
+  maxRetries = 5;
+
+  private queue = new Map<string, Queue>();
+
+  constructor(options?: HTTPClientOptions) {
+    if (options !== undefined) {
+      if (
+        (options.token !== undefined && options.tokenType === undefined) ||
+        (options.token === undefined && options.tokenType !== undefined)
+      ) {
+        throw new TypeError(
+          `Either tokenType and token both must be specified or neither of should be.`,
+        );
+      }
+    }
+    this.tokenType = options?.tokenType;
+    this.token = options?.token;
+    this.userAgent = options?.userAgent ?? USER_AGENT;
+    this.baseURL = options?.baseURL ?? DISCORD_API_BASE;
+    this.version = options?.version ?? DISCORD_API_VERSION;
+    this.maxRetries = options?.maxRetries ?? this.maxRetries;
+  }
+
+  async request(
+    method: HTTPMethod,
+    endpoint: Endpoint,
+    options?: RequestOptions,
+  ) {
+    const bucket = options?.bucket ?? getRouteBucket(endpoint);
+    const queue = this.queue.get(bucket) ??
+      (this.queue.set(bucket, new Queue()), this.queue.get(bucket)!);
+
+    const execute = async (): Promise<any> => {
```

Comment:
If possible remove the `any` type so that it is inferred

Reply:
yeah we can use generic i think